### PR TITLE
Experimental/sf

### PR DESCRIPTION
**Pull request by lqdung-lockon**

### CODE REVIEW

File: Event/Event.php [hunk 1]
```diff
@@ -10,40 +10,77 @@
...
 use Eccube\Event\EventArgs;
+use Eccube\Common\EccubeConfig;
+use Eccube\Repository\Master\ProductStatusRepository;
+use Eccube\Entity\Master\ProductStatus;
+use Symfony\Component\Form\Extension\Core\Type\CollectionType;
+use Plugin\RelatedProduct\Form\Type\Admin\RelatedProductType;
+use Doctrine\ORM\EntityManagerInterface;
 
 /**
  * Class Event for  new hook point on version >= 3.0.9.
```
Comment (lqdung-lockon) [inline Event/Event.php]: remove this comment

File: Controller/Admin/RelatedProductController.php [hunk 2]
```diff
@@ -12,35 +12,80 @@
 
 use Eccube\Application;
```
Comment (lqdung-lockon) [inline Controller/Admin/RelatedProductController.php]: remove this

File: Controller/Admin/RelatedProductController.php [hunk 3]
```diff
@@ -12,35 +12,80 @@
...
 use Symfony\Component\HttpFoundation\Request;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method;
```
Comment (lqdung-lockon) [inline Controller/Admin/RelatedProductController.php]: remove this

File: Event/Event.php [hunk 4]
```diff
@@ -110,14 +154,14 @@ public function onRenderAdminProductInit(EventArgs $event)
         /** @var FormBuilder $builder */
         $builder = $event->getArgument('builder');
         $builder
-            ->add('related_collection', 'collection', array(
+            ->add('related_collection', CollectionType::class, [
                 'label' => '関連商品',
```
Comment (lqdung-lockon) [inline Event/Event.php]: use message translator

File: Event/Event.php [hunk 5]
```diff
@@ -131,21 +175,29 @@ public function onRenderAdminProductInit(EventArgs $event)
...
+            $modal = $modal->getCode();
+        } catch (\Exception $e) {
+            $snipet = '';
+            $modal = '';
+        }
 
         //add related product to product edit
-        $search = '<div id="detail_box__footer" class="row hidden-xs hidden-sm">';
+        // TODO: should find a way better to insert/edit template
+        // TODO: @admin/Product/product.twig should have a anchor (id, or comment tag) to insert/edit template
```
Comment (lqdung-lockon) [inline Event/Event.php:192]: Current, due to don't have `<!-- layout.admin.product.primaryCol.footer -->`. so please add to final block of  `<div class="card rounded border-0 mb-4">` or something like that

File: Event/Event.php [hunk 6]
```diff
@@ -162,23 +214,26 @@ public function onRenderAdminProduct(TemplateEvent $event)
...
+            $form = $event->getArgument('form');
+            $this->relatedProductRepository->removeChildProduct($Product);
+            log_info('remove all now related product data of ', ['Product id' => $Product->getId()]);
+            $RelatedProducts = $form->get('related_collection')->getData();
+            foreach ($RelatedProducts as $RelatedProduct) {
+                /* @var $RelatedProduct \Plugin\RelatedProduct\Entity\RelatedProduct */
+                if ($RelatedProduct->getChildProduct() instanceof Product) {
+                    $RelatedProduct->setProduct($Product);
+                    $this->entityManager->persist($RelatedProduct);
+                    $this->entityManager->flush();
```
Comment (lqdung-lockon) [inline Event/Event.php]: Please flush a definite object (policy of plugin)
Other places are similar.

File: Repository/RelatedProductRepository.php [hunk 7]
```diff
@@ -10,46 +10,62 @@
...
      */
     public function removeChildProduct($Product)
     {
         $em = $this->getEntityManager();
-        $Children = $this->findBy(array('Product' => $Product));
+        $Children = $this->findBy(['Product' => $Product]);
         foreach ($Children as $Child) {
             $em->remove($Child);
-            $em->flush($Child);
+            $em->flush();
```
Comment (lqdung-lockon) [inline Repository/RelatedProductRepository.php]: Same above...

File: EventSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * Copyright (C) 2016 LOCKON CO.,LTD. All Rights Reserved.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+*/
+
+namespace Plugin\RelatedProduct;
+
+use Eccube\Application;
```
Comment (lqdung-lockon) [inline EventSubscriber.php]: remove un-use code

**Comment by lqdung-lockon** [comment]
image

- Don't show image
- button "商品を選択" and "削除" don't same style.

**Comment by lqdung-lockon** [comment]
image

don't see the icon next/back

**Comment by lqdung-lockon** [comment]
image
broken layout

File: Event/Event.php (hunk 1 above)
Comment (ndquocphong) [reply Event/Event.php]: @lqdung-lockon @toannguyen-lockon @hoand-vn done

File: Controller/Admin/RelatedProductController.php (hunk 3 above)
Comment (ndquocphong) [reply Controller/Admin/RelatedProductController.php]: @lqdung-lockon @toannguyen-lockon @hoand-vn done

File: Event/Event.php (hunk 4 above)
Comment (ndquocphong) [reply Event/Event.php]: @lqdung-lockon @toannguyen-lockon @hoand-vn done

File: EventSubscriber.php (hunk 8 above)
Comment (ndquocphong) [reply EventSubscriber.php]: @lqdung-lockon @toannguyen-lockon @hoand-vn done

File: Event/Event.php (hunk 6 above)
Comment (ndquocphong) [reply Event/Event.php]: @lqdung-lockon @toannguyen-lockon @hoand-vn done

File: Repository/RelatedProductRepository.php (hunk 7 above)
Comment (ndquocphong) [reply Repository/RelatedProductRepository.php]: @lqdung-lockon @toannguyen-lockon @hoand-vn done

File: Event/Event.php (hunk 5 above)
Comment (ndquocphong) [reply Event/Event.php:192]: @lqdung-lockon @toannguyen-lockon @hoand-vn 
> please add to final block of

How to determine the final block `<div class="card rounded border-0 mb-4">` ? 


> or something like that

Please suggest

**Comment by ndquocphong** [comment]
@lqdung-lockon @toannguyen-lockon @hoand-vn

Layout broken because the related product block inject into new style guide DOM inccorrect way. I have ask a question on sheet Q/A for this situation

**Comment by hoand-vn** [comment]
finish  fixe

File: Event/Event.php (hunk 5 above)
Comment (lqdung-lockon) [reply Event/Event.php:192]: for example: https://github.com/eccubevn/coupon-plugin/blob/experimental/sf/Event.php#L229

Comment (ndquocphong) [reply Event/Event.php:192]: @lqdung-lockon thank you for suggest, I noted, will fix after have feedback from JP

**Pull request closed without merging**